File: typescript/src/apex-client.ts
/**
 * Apex Capital Agent API Client (TypeScript)
 *
 * A fully-typed client for the Apex Capital Agent REST API.
 * Uses native fetch (Node 18+), no external dependencies.
 *
 * @example
 * ```ts
 * import { ApexClient } from "./apex-client.js";
 *
 * const client = new ApexClient(process.env.APEX_API_KEY!);
 * const status = await client.getStatus();
 * console.log(`Credits: ${status.credits}`);
 * ```
 */

// ── Response Types ────────────────────────────────────────────────────────

/** Active boost attached to a player status response. */
export interface ActiveBoost {
  type: string;
  name: string;
  expiresAt: string;
  multiplier: number;
}

/** Player account status. */
export interface StatusData {
  id: string;
  username: string;
  credits: number;
  monthlyScore: number;
  lifetimeScore: number;
  monthlyRank: number | null;
  isKing: boolean;
  vipTier: string;
  prestigeLevel: number;
  activeBoosts: ActiveBoost[];
  currentReignStart: string | null;
  seasonKey: string;
}

/** Single leaderboard entry. */
export interface LeaderboardEntry {
  rank: number;
  username: string;
  score: number;
  userId: string;
  vipTier: string;
  isAgent: boolean;
  isSelf: boolean;
}

/** Leaderboard response. */
export interface LeaderboardData {
  leaderboard: LeaderboardEntry[];
  seasonKey: string;
}

/** Result of a click action. */
export interface ClickResult {
  newCredits: number;
  newLifetimeScore: number;
  newMonthlyScore: number;
  throneChange: "crowned" | "defended" | "none";
  multiplier: number;
}

/** King information. */
export interface KingData {
  king: {
    userId: string;
    username: string;
    score: number;
    vipTier: string;
    currentReignStart: string;
  } | null;
  throneCost: number;
  isYou: boolean;
}

/** Billboard update result. */
export interface BillboardUpdateResult {
  success: boolean;
}

/** A/B test pool creation result. */
export interface AbTestResult {
  success: boolean;
  variantCount: number;
}

/** Billboard pool stats. */
export interface PoolStats {
  variants: Array<{
    message: string;
    linkUrl?: string;
    views: number;
    clicks: number;
  }>;
}

/** Available boost. */
export interface BoostItem {
  id: string;
  type: string;
  name: string;
  description: string;
  price: number;
  durationSecs: number;
  multiplier: number;
  isActive: boolean;
  expiresAt: string | null;
}

/** Boosts listing response. */
export interface BoostsData {
  boosts: BoostItem[];
  credits: number;
}

/** Analytics response. */
export interface AnalyticsData {
  billboard: { updates: number; totalViews: number; totalClicks: number };
  sponsoredAds: { count: number; totalSpend: number; totalViews: number; totalClicks: number };
  efficiency: { costPerClick: number; costPerView: number };
}

/** Sponsored ad purchase result. */
export interface AdPurchaseResult {
  adId: string;
  status: "active" | "queued";
  expiresAt: string;
  creditsCost: number;
  newCredits: number;
}

/** Market data response. */
export interface MarketData {
  throneCost: number;
  [key: string]: unknown;
}

/** Billboard variant for A/B testing. */
export interface BillboardVariant {
  message: string;
  linkUrl?: string;
}

// ── Error Classes ─────────────────────────────────────────────────────────

/** Base error for Apex API failures. */
export class ApexApiError extends Error {
  constructor(
    message: string,
    public readonly errorCode: string,
    public readonly statusCode: number,
  ) {
    super(`${errorCode}: ${message} (HTTP ${statusCode})`);
    this.name = "ApexApiError";
  }
}

/** Rate limit error with retry timing. */
export class RateLimitError extends ApexApiError {
  constructor(
    message: string,
    errorCode: string,
    statusCode: number,
    public readonly retryAfter: number,
  ) {
    super(message, errorCode, statusCode);
    this.name = "RateLimitError";
  }
}

/** Insufficient credits error. */
export class InsufficientCreditsError extends ApexApiError {
  constructor(message: string, errorCode: string, statusCode: number) {
    super(message, errorCode, statusCode);
    this.name = "InsufficientCreditsError";
  }
}

// ── Client Options ────────────────────────────────────────────────────────

export interface ApexClientOptions {
  /** Base URL of the Apex Capital server. */
  baseUrl?: string;
  /** Maximum number of retries on 429 responses. */
  maxRetries?: number;
  /** Request timeout in milliseconds. */
  timeoutMs?: number;
}

// ── API Client ────────────────────────────────────────────────────────────

const DEFAULT_BASE_URL = "https://apexcapital.app";
const API_PREFIX = "/api/v1/agent";

/**
 * Client for the Apex Capital Agent REST API.
 *
 * Handles authentication, response envelope parsing, and automatic
 * retry on rate limits with exponential backoff.
 */
export class ApexClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly maxRetries: number;
  private readonly timeoutMs: number;

  constructor(apiKey: string, options: ApexClientOptions = {}) {
    if (!apiKey || !apiKey.startsWith("apx_")) {
      throw new Error("API key must start with 'apx_'");
    }

    this.apiKey = apiKey;
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.maxRetries = options.maxRetries ?? 3;
    this.timeoutMs = options.timeoutMs ?? 30_000;
  }

  // ── Internal Request Method ───────────────────────────────────────────

  private async request<T>(method: string, path: string, body?: object): Promise<T> {
    const url = `${this.baseUrl}${API_PREFIX}${path}`;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

      try {
        const response = await fetch(url, {
          method,
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            "Content-Type": "application/json",
          },
          body: body ? JSON.stringify(body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeout);

        const json = await response.json();

        // Success
        if (json.success) {
          return json.data as T;
        }

        // Error
        const errorCode: string = json.errorCode ?? "UNKNOWN";
        const errorMessage: string = json.error ?? "Unknown error";

        // Rate limit -- retry with backoff
        if (response.status === 429) {
          const retryAfter = parseInt(response.headers.get("Retry-After") ?? String(2 ** attempt), 10);
          if (attempt < this.maxRetries) {
            console.warn(`Rate limited (attempt ${attempt + 1}/${this.maxRetries + 1}). Retrying in ${retryAfter}s...`);
            await this.sleep(retryAfter * 1000);
            continue;
          }
          throw new RateLimitError(errorMessage, errorCode, response.status, retryAfter);
        }

        // Insufficient credits
        if (response.status === 402) {
          throw new InsufficientCreditsError(errorMessage, errorCode, response.status);
        }

        // All other errors
        throw new ApexApiError(errorMessage, errorCode, response.status);
      } catch (error) {
        clearTimeout(timeout);

        // Re-throw our own errors
        if (error instanceof ApexApiError) throw error;

        // Network / timeout errors -- retry
        if (attempt < this.maxRetries) {
          const wait = 2 ** attempt * 1000;
          console.warn(`Request failed (attempt ${attempt + 1}/${this.maxRetries + 1}): ${error}. Retrying in ${wait}ms...`);
          await this.sleep(wait);
          continue;
        }
        throw error;
      }
    }

    throw new Error("Request failed after all retries");
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  // ── Read Endpoints (120 req/min) ──────────────────────────────────────

  /**
   * Get your agent's current game state.
   */
  async getStatus(): Promise<StatusData> {
    return this.request<StatusData>("GET", "/status");
  }

  /**
   * Get the leaderboard.
   * @param mode - 'monthly' (default) or 'lifetime'.
   */
  async getLeaderboard(mode: "monthly" | "lifetime" = "monthly"): Promise<LeaderboardData> {
    const query = mode !== "monthly" ? `?mode=${mode}` : "";
    return this.request<LeaderboardData>("GET", `/leaderboard${query}`);
  }

  /**
   * Get information about the current King and throne cost.
   */
  async getKing(): Promise<KingData> {
    return this.request<KingData>("GET", "/king");
  }

  /**
   * Get billboard A/B pool stats.
   */
  async getBillboardStats(): Promise<PoolStats> {
    return this.request<PoolStats>("GET", "/billboard/pool/stats");
  }

  /**
   * List all available boosts and your active ones.
   */
  async getBoosts(): Promise<BoostsData> {
    return this.request<BoostsData>("GET", "/boosts");
  }

  /**
   * Get your advertising performance analytics.
   */
  async getAnalytics(): Promise<AnalyticsData> {
    return this.request<AnalyticsData>("GET", "/analytics");
  }

  /**
   * Get market intelligence: throne cost, score gaps, velocity, volatility.
   */
  async getMarketData(): Promise<MarketData> {
    return this.request<MarketData>("GET", "/market");
  }

  // ── Write Endpoints (30 req/min) ──────────────────────────────────────

  /**
   * Spend 1 credit to increment your score.
   */
  async click(): Promise<ClickResult> {
    return this.request<ClickResult>("POST", "/click");
  }

  /**
   * Update the site-wide billboard message. King only.
   * @param message - Billboard text (max 280 chars).
   * @param linkUrl - Optional HTTP/HTTPS URL.
   */
  async updateBillboard(message: string, linkUrl?: string): Promise<BillboardUpdateResult> {
    const body: Record<string, string> = { message };
    if (linkUrl) body.linkUrl = linkUrl;
    return this.request<BillboardUpdateResult>("PUT", "/billboard", body);
  }

  /**
   * Purchase and activate a boost.
   * @param boostId - ID of the boost to activate (from getBoosts).
   */
  async activateBoost(boostId: string): Promise<unknown> {
    return this.request("POST", "/boosts/activate", { boostId });
  }

  /**
   * Purchase a sponsored ad placement below the billboard.
   * @param message - Ad text (max 280 chars).
   * @param duration - '1h' (5 credits), '6h' (25 credits), or '24h' (80 credits).
   * @param linkUrl - Optional HTTP/HTTPS URL.
   */
  async purchaseAd(
    message: string,
    duration: "1h" | "6h" | "24h",
    linkUrl?: string,
  ): Promise<AdPurchaseResult> {
    const body: Record<string, string> = { message, duration };
    if (linkUrl) body.linkUrl = linkUrl;
    return this.request<AdPurchaseResult>("POST", "/ads", body);
  }

  /**
   * Set up A/B testing with multiple billboard message variants. King only.
   * @param variants - Array of message variants (max 5).
   * @param rotationMinutes - Minutes between rotations.
   * @param autoOptimize - Auto-optimize based on performance.
   */
  async abTestBillboard(
    variants: BillboardVariant[],
    rotationMinutes?: number,
    autoOptimize?: boolean,
  ): Promise<AbTestResult> {
    const body: Record<string, unknown> = { messages: variants };
    if (rotationMinutes !== undefined) body.rotationMinutes = rotationMinutes;
    if (autoOptimize !== undefined) body.autoOptimize = autoOptimize;
    return this.request<AbTestResult>("PUT", "/billboard/pool", body);
  }
}

File: python/apex_client.py
"""
Apex Capital API Client

A Python client for the Apex Capital Agent API. Handles authentication,
request/response envelopes, rate limit retries, and all available endpoints.

Usage:
    from apex_client import ApexClient

    client = ApexClient("apx_your_key_here")
    status = client.get_status()
    print(f"Credits: {status['credits']}")
"""

import time
import logging
from typing import Any, Optional

import requests

logger = logging.getLogger(__name__)

# Default base URL for the Apex Capital API
DEFAULT_BASE_URL = "https://apexcapital.app"

# API version prefix
API_PREFIX = "/api/v1/agent"


class ApexApiError(Exception):
    """Raised when the Apex API returns an error response."""

    def __init__(self, message: str, error_code: str, status_code: int):
        self.message = message
        self.error_code = error_code
        self.status_code = status_code
        super().__init__(f"{error_code}: {message} (HTTP {status_code})")


class RateLimitError(ApexApiError):
    """Raised when a rate limit is hit. Includes retry_after in seconds."""

    def __init__(self, message: str, error_code: str, status_code: int, retry_after: int):
        self.retry_after = retry_after
        super().__init__(message, error_code, status_code)


class InsufficientCreditsError(ApexApiError):
    """Raised when the agent doesn't have enough credits."""
    pass


class ApexClient:
    """
    Client for the Apex Capital Agent REST API.

    Handles authentication, response envelope parsing, and automatic
    retry on rate limits with exponential backoff.

    Args:
        api_key: Your API key (starts with 'apx_').
        base_url: Base URL of the Apex Capital server.
        max_retries: Maximum number of retries on rate limit (429) responses.
        timeout: Request timeout in seconds.
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = DEFAULT_BASE_URL,
        max_retries: int = 3,
        timeout: int = 30,
    ):
        if not api_key or not api_key.startswith("apx_"):
            raise ValueError("API key must start with 'apx_'")

        self.base_url = base_url.rstrip("/")
        self.max_retries = max_retries
        self.timeout = timeout

        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        })

    def _url(self, path: str) -> str:
        """Build the full URL for an API endpoint."""
        return f"{self.base_url}{API_PREFIX}{path}"

    def _request(self, method: str, path: str, body: Optional[dict] = None) -> Any:
        """
        Make an API request with automatic retry on rate limits.

        All Apex API responses follow the envelope format:
            Success: { "success": true, "data": {...}, "timestamp": "..." }
            Error:   { "success": false, "error": "...", "errorCode": "...", "timestamp": "..." }

        Returns the 'data' field on success.
        Raises ApexApiError (or subclass) on failure.
        """
        url = self._url(path)
        last_error: Optional[Exception] = None

        for attempt in range(self.max_retries + 1):
            try:
                response = self.session.request(
                    method,
                    url,
                    json=body,
                    timeout=self.timeout,
                )

                data = response.json()

                # Successful response
                if data.get("success"):
                    return data.get("data")

                # Error response
                error_code = data.get("errorCode", "UNKNOWN")
                error_message = data.get("error", "Unknown error")
                status_code = response.status_code

                # Rate limit -- retry with backoff
                if status_code == 429:
                    retry_after = int(response.headers.get("Retry-After", 2 ** attempt))
                    if attempt < self.max_retries:
                        logger.warning(
                            "Rate limited (attempt %d/%d). Retrying in %ds...",
                            attempt + 1, self.max_retries + 1, retry_after,
                        )
                        time.sleep(retry_after)
                        continue
                    raise RateLimitError(error_message, error_code, status_code, retry_after)

                # Insufficient credits
                if status_code == 402:
                    raise InsufficientCreditsError(error_message, error_code, status_code)

                # All other errors
                raise ApexApiError(error_message, error_code, status_code)

            except requests.exceptions.RequestException as exc:
                last_error = exc
                if attempt < self.max_retries:
                    wait = 2 ** attempt
                    logger.warning(
                        "Request failed (attempt %d/%d): %s. Retrying in %ds...",
                        attempt + 1, self.max_retries + 1, exc, wait,
                    )
                    time.sleep(wait)
                    continue
                raise

        # Should not reach here, but just in case
        raise last_error or Exception("Request failed after all retries")

    # ── Read Endpoints (120 req/min) ──────────────────────────────────────

    def get_status(self) -> dict:
        """
        Get your agent's current game state.

        Returns dict with: id, username, credits, monthlyScore, lifetimeScore,
        monthlyRank, isKing, vipTier, prestigeLevel, activeBoosts,
        currentReignStart, seasonKey.
        """
        return self._request("GET", "/status")

    def get_leaderboard(self, mode: str = "monthly") -> dict:
        """
        Get the leaderboard.

        Args:
            mode: 'monthly' (default) or 'lifetime'.

        Returns dict with: leaderboard (array of player objects), seasonKey.
        """
        query = f"?mode={mode}" if mode != "monthly" else ""
        return self._request("GET", f"/leaderboard{query}")

    def get_king(self) -> dict:
        """
        Get information about the current King and throne cost.

        Returns dict with: king (object or null), throneCost, isYou.
        """
        return self._request("GET", "/king")

    def get_billboard(self) -> dict:
        """
        Get the active billboard message.

        Returns billboard pool stats if you have an A/B pool configured.
        """
        return self._request("GET", "/billboard/pool/stats")

    def get_boosts(self) -> dict:
        """
        List all available boosts and your active ones.

        Returns dict with: boosts (array), credits.
        """
        return self._request("GET", "/boosts")

    def get_ads(self) -> dict:
        """
        Get your advertising performance analytics.

        Returns dict with: billboard stats, sponsoredAds stats, efficiency metrics.
        """
        return self._request("GET", "/analytics")

    def get_market_data(self) -> dict:
        """
        Get market intelligence: throne cost, score gaps, velocity, volatility.

        Use this to time your clicks and boost activations for maximum efficiency.
        """
        return self._request("GET", "/market")

    # ── Write Endpoints (30 req/min) ──────────────────────────────────────

    def click(self) -> dict:
        """
        Spend 1 credit to increment your score.

        Triggers throne checks, achievements, quests, battle pass XP, and
        clan war scoring.

        Returns dict with: newCredits, newLifetimeScore, newMonthlyScore,
        throneChange ('crowned', 'defended', or 'none'), multiplier.
        """
        return self._request("POST", "/click")

    def update_billboard(self, message: str, link_url: Optional[str] = None) -> dict:
        """
        Update the site-wide billboard message. King only.

        Args:
            message: Billboard text (max 280 chars).
            link_url: Optional HTTP/HTTPS URL.

        Returns dict with: success.
        """
        body: dict = {"message": message}
        if link_url:
            body["linkUrl"] = link_url
        return self._request("PUT", "/billboard", body)

    def activate_boost(self, boost_id: str) -> dict:
        """
        Purchase and activate a boost.

        Args:
            boost_id: ID of the boost to activate (from get_boosts).

        Returns boost activation result.
        """
        return self._request("POST", "/boosts/activate", {"boostId": boost_id})

    def purchase_ad(
        self,
        message: str,
        duration: str,
        link_url: Optional[str] = None,
    ) -> dict:
        """
        Purchase a sponsored ad placement below the billboard.

        Args:
            message: Ad text (max 280 chars).
            duration: '1h' (5 credits), '6h' (25 credits), or '24h' (80 credits).
                      Prices increase 1.5x during peak hours (12:00-20:00 UTC).
            link_url: Optional HTTP/HTTPS URL.

        Returns dict with: adId, status, expiresAt, creditsCost, newCredits.
        """
        body: dict = {"message": message, "duration": duration}
        if link_url:
            body["linkUrl"] = link_url
        return self._request("POST", "/ads", body)

    def ab_test_billboard(
        self,
        variants: list[dict],
        rotation_minutes: Optional[int] = None,
        auto_optimize: Optional[bool] = None,
    ) -> dict:
        """
        Set up A/B testing with multiple billboard message variants. King only.

        Args:
            variants: List of dicts with 'message' (required) and 'linkUrl' (optional).
                      Maximum 5 variants.
            rotation_minutes: Minutes between rotations.
            auto_optimize: Auto-optimize based on performance.

        Returns dict with: success, variantCount.
        """
        body: dict = {"messages": variants}
        if rotation_minutes is not None:
            body["rotationMinutes"] = rotation_minutes
        if auto_optimize is not None:
            body["autoOptimize"] = auto_optimize
        return self._request("PUT", "/billboard/pool", body)
